### DISCUSSION TITLE
Best way to add custom autocomplete/annotations?

### DISCUSSION THREAD

**Discussion by gingerbeardman** [zed-industries/zed#6643] [Feature Requests]
I'm a game developer, and use an SDK with functions written in Lua.

I managed to get autocompletition/hints to pick up the Playdate function definitions after reading:
- https://github.com/LuaLS/lua-language-server/wiki/Annotations
- https://github.com/LuaLS/lua-language-server/wiki/Libraries

To do so I use the stub.lua and .luarc.json from either of
- https://github.com/notpeter/playdate-luacats (detailed)
- https://github.com/notpeter/playdate-docdef (minimal)

...and add them to my project folder.

Anyway, I'm new to Zed so I'm wondering if this is the best way?

> I tried adding it at:
> - `~/Library/Application Support/Zed/languages/lua-language-server/meta/playdate/stub.lua`
> ...but this did not work.

Minimal
![Screen shot 2023-12-10 at 15 45 25](https://github.com/zed-industries/community/assets/49612/1e11ca68-a073-46dd-b339-f039ff8933cc)

Detailed
![Screen shot 2023-12-10 at 15 53 57](https://github.com/zed-industries/community/assets/49612/e4924da0-1922-4c42-a4b9-65f9d2317ed2)

I can modify the stub.lua to markdown format all URLs to make them clickable in the annotation.

![Screen shot 2023-12-10 at 15 57 03](https://github.com/zed-industries/community/assets/49612/cd63c2b0-bf58-4960-83f8-86ab69679a6e)

**Comment by Stanzilla** [comment]
I'm also looking for a way to do this, in VSC, the extension supports a workspace based library setting that you can feed files/folders to, an equivalent to this in Zed would be great.

**Reply by and3rson** [reply]
Looking for the same thing. We've integrated Lua into our community-driven DIY console project (https://github.com/and3rson/lilka) and our library defines a lot of global functions for graphics, inputs, hardware, etc.
I'm wondering what's the best way to give our users auto-completion for writing Lua scripts for our platform.

**Comment by notpeter** [comment]
@Stanzilla The editor-agnostic way to configure LuaLS is with `.luarc.json`, it supports the same keys you're using in your VSCode workspace config but without the `lua.` prefix. This works for me in Zed.

**Reply by Stanzilla** [reply]
It did not work at the time of my posting but I can test again.

**Reply by notpeter** [reply]
Curious if you were ever able to get this working and would love to help troubleshoot if not.

**Reply by Stanzilla** [reply]
> Curious if you were ever able to get this working and would love to help troubleshoot if not.

I haven't tried Lua in Zed since then, but I do have a repo with a luarc.json file https://github.com/WeakAuras/WeakAuras2

**Comment by notpeter** [comment]
I think a per-project `.luarc.json` is the right answer here.  It supports the same features as the `lua.*` namespace items (like `lua.workspace` in `.vscode/settings.json`) as the sumneko.lua VSCode Lua extension just without the `lua` prefix.

Notably LuaLS supports `.luarc.json` universally across editors (VSCode, NeoVim, Zed, etc) without requiring extension specific settings.

The second question is how to have your project find external stubs:

1. Use an absolute path in `workspace` pointing at wherever you checked out the stubs (simple and works great for solo playdate developers)
1. Use a relative path in `workspace` and include instructions to checkout the stubs repo in a peer directory of your project (e.g. `~/code/playdate-luacats` and `~/code/projectname` -- and `"workspace": ["../playdate-luacats"]`)
1. Use git submodules in your repo (ideal, but more complex for users)

@gingerbeardman I changed the format of links in annotations to use markdown to make them clickable in Zed in response to the above, thanks for catching that. Commit here: https://github.com/notpeter/playdate-luacats/commit/98861f14179da95448c4f32577712a6df24303e5

The above is focused on Lua autocomplete/annotations in Zed via LuaLS, but if you wanted to move beyond that you Zed supports code intelligence (autocomplete, snippets, linting, etc) custom language servers and to a lesser extent via tree-sitter grammars.

**Reply by jpike88** [reply]
A project-level snippet config, like the ability to point to a file for a given language (so we can just rely on a npm module instead), would be a big deal. This would replace how we used es6 javascript snippet extension as recommended extensions so all devs had the same workflow enabled. (though having recommended extensions would rock as well)

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/community-links
